### PR TITLE
Adding OSS Redis based Historical Statistics Provider

### PR DESCRIPTION
**Pull request by jaystarshot**
## Description
Adding an Open source Redis based Historical Statistics Provider.
The Redis client is stateful and based on Lettuce [link](https://github.com/lettuce-io/lettuce-core).
There are proper timeouts for fetch/get requests which are  configurable.
Latency and other metrics are available and exported thorough the JMX exporter. 
Redis requests will be pipelined since Lettuce uses [Pipelining](https://redis.io/docs/manual/pipelining/) by default

The provider plugin would need to be configured based on the use case, docs added - [here](presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst)

## Motivation and Context
There is no proper oss implementation of a statistics provider and this PR adds one

## Test Plan
Unit tests + production tests

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Adding an open source redis based historical statistics provider

### CODE REVIEW

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        if (this.propertyMap == null) {
+            log.info("Redis Provider Plugin returned an empty instance");
+            return ImmutableList.of(EmptyPlanStatisticsProvider.getInstance());
+        }
+        if (!(propertyMap.getOrDefault(RedisProviderConfig.COORDINATOR_PROPERTY_NAME, "false").equals("true") &&
+                propertyMap.getOrDefault(RedisProviderConfig.HBO_PROVIDER_ENABLED_NAME, "false").equals("true"))) {
+            return ImmutableList.of(EmptyPlanStatisticsProvider.getInstance());
+        }
+        Injector injector = intializeAndGetInjector();
```
Comment (jaystarshot) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java:69]: Technically this could be done in constructor but I think doing in getHistoryBasedPlanStatisticsProviders makes more sense since this is where the provider is loaded at runtime

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
+=======================
+Redis Hbo Provider
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:2]:
```suggestion
Redis HBO Provider
```

I would keep capitalization the same in the docs, all caps because it's an acronym.

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+=======================
+
+Redis Hbo Provider allows loading a custom configured Redis Client for storing/retrieving Historical Stats. The Redis client is stateful and is based on
+Lettuce https://github.com/lettuce-io/lettuce-core
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:6]: Hyperlink?

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ <dependency>
+            <groupId>com.facebook.presto</groupId>
+            <artifactId>presto-hbo-provider</artifactId>
+            <version>${project.version}</version>
+            <type>zip</type>
+            <scope>provided</scope>
+ </dependency>
+``
+
+4. Add a custom redis client in src/main/java/com/facebook/presto/statistic/RedisClusterClientProvider.java. Please note that a regular RedisClient might be sufficient; it's not obligatory to utilize a RedisClusterClient.
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:61]: Why would one need a custom redis client?

File: presto-server/pom.xml [hunk 5]
```diff
@@ -316,6 +316,7 @@
             <scope>provided</scope>
         </dependency>
 
+
```
Comment (tdcmeehan) [inline presto-server/pom.xml:319]: Unrelated change

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    @Override
+    public ByteBuffer encodeKey(String key)
+    {
+        return ByteBuffer.wrap(key.getBytes(StandardCharsets.UTF_8));
+    }
+
+    @Override
+    public ByteBuffer encodeValue(HistoricalPlanStatistics historicalPlanStatistics)
+    {
+        ThriftCodec<HistoricalPlanStatistics> writeCodec = thriftCodecManager.getCodec(HistoricalPlanStatistics.class);
```
Comment (tdcmeehan) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java:59]: Extract this out into a member variable to cache?

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Redis codec implementation for string keys and HistoricalPlanStatistics values.
+ */
+public class HistoricalStatisticsSerde
+        implements RedisCodec<String, HistoricalPlanStatistics>
+{
+    private static final int estimatedBuffersize = 100 * 1024;
```
Comment (tdcmeehan) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java:36]:
```suggestion
    private static final int ESTIMATED_BUFFER_SIZE = 100 * 1024;
```

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisClusterClientProvider.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import io.lettuce.core.cluster.RedisClusterClient;
+
+import java.util.Map;
+
+public class RedisClusterClientProvider
+{
+    private RedisClusterClientProvider()
+    {
+    }
+    public static RedisClusterClient get(Map<String, String> properties)
```
Comment (tdcmeehan) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisClusterClientProvider.java:25]:
```suggestion
    private RedisClusterClientProvider() {}
    
    public static RedisClusterClient get(Map<String, String> properties)
```

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Please also make sure the commits follow our guidelines.  In particular, the commit message guidelines, and remember to squash your commits (see CONTRIBUTING.md).

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+Configuration properties
+------------------------
+
+The following configuration properties are available:
+
+===================================     =====================================================================
+Property Name                           Description
+===================================     =====================================================================
+``coordinator``                         Boolean property whether presto server a coordinator
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:21]: perhaps "whether Presto server is a coordinator"

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst (hunk 2 above)
Comment (steveburnett) [reply presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:2]: The acronym HBO occurs in more places than this line - line 5 and line 11 as of this comment and perhaps elsewhere - so I would suggest changing this throughout the page for consistency.

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+``hbo.provider.total-set-timeoutms``    Maximum timeout in ms for redis set requests
+``hbo.provider.default-ttl-sec``        TTL in sec of the redis data to be stored
+``hbo.provider.enabled``                Boolean property whether this plugin in enabled in production
+``credentials-path``                    Path for redis credentials
+===================================     =====================================================================
+
+Credentials
+-----------
+
+The connector requires the redis server server uri in order to access Redis ``hbo.provider.server_uri``
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:33]: "redis server server uri": "server" is repeated, possibly unintentionally. Could you use the name "Redis Server URI" that you gave this in line 22?

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+===================================     =====================================================================
+Property Name                           Description
+===================================     =====================================================================
+``coordinator``                         Boolean property whether presto server a coordinator
+``hbo.provider.server_uri``             Redis Server URI
+``hbo.provider.total-fetch-timeoutms``  Maximum timeout in ms for redis fetch requests
+``hbo.provider.total-set-timeoutms``    Maximum timeout in ms for redis set requests
+``hbo.provider.default-ttl-sec``        TTL in sec of the redis data to be stored
+``hbo.provider.enabled``                Boolean property whether this plugin in enabled in production
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:26]: perhaps "Boolean property whether this plugin is enabled in production"

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+The following configuration properties are available:
+
+===================================     =====================================================================
+Property Name                           Description
+===================================     =====================================================================
+``coordinator``                         Boolean property whether presto server a coordinator
+``hbo.provider.server_uri``             Redis Server URI
+``hbo.provider.total-fetch-timeoutms``  Maximum timeout in ms for redis fetch requests
+``hbo.provider.total-set-timeoutms``    Maximum timeout in ms for redis set requests
+``hbo.provider.default-ttl-sec``        TTL in sec of the redis data to be stored
```
Comment (feilong-liu) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:25]: default-ttl-sec -> default-ttl-seconds
TTL in sec of ... -> TTL in seconds of ...

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java (hunk 7 above)
Comment (feilong-liu) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java:36]: I assume this is bytes? Consider renaming to ESTIMATED_BUFFER_SIZE_BYTES?

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java (hunk 6 above)
Comment (feilong-liu) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java:59]: Is it possible to have this as a static member variable? Similar for the readCodec below

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisPlanStatisticsProvider.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    private final RedisProviderApiStats redisProviderApiStats;
+    private RedisClusterAsyncCommands<String, HistoricalPlanStatistics> commands;
+
+    private final long totalFetchTimeoutMillis;
+
+    private final long totalSetTimeMillis;
+
+    private final long defaultTTLSeconds;
+
+    public RedisPlanStatisticsProvider(RedisClusterAsyncCommands<String, HistoricalPlanStatistics> commands, long totalFetchTimeoutMillis, long totalSetTimeoutMillis, long defaultTTLSeconds)
```
Comment (feilong-liu) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisPlanStatisticsProvider.java:46]: @Inject ?

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+            propertyMap = loadAndProcessProperties();
+        }
+        catch (NoSuchFileException fileNotFoundException) {
+            propertyMap = null;
+            log.info("Redis Provider Plugin was not loaded since config file not found");
+        }
+        this.propertyMap = propertyMap;
+    }
+
+    public Iterable<HistoryBasedPlanStatisticsProvider> getHistoryBasedPlanStatisticsProviders()
```
Comment (feilong-liu) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java:59]: @Override

Comment (feilong-liu) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java:59]: Is it cleaner if we have a factory method to create the provider instance?

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst (hunk 4 above)
Comment (jaystarshot) [reply presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:61]: There can be multiple redis deployment strategies - [eg](https://medium.com/hepsiburadatech/redis-solutions-standalone-vs-sentinel-vs-cluster-f46e703307a9) Standalone, sentinal or cluster. 

I think lettuce supports Standalone and sentinal deployments with [RedisClient](https://github.com/lettuce-io/lettuce-core/blob/bfb6880afb4fbb70110f5cd8252af69806a104d9/src/main/java/io/lettuce/core/RedisClient.java#L89).

RedisClusterClient with [RedisClusterClient](https://github.com/lettuce-io/lettuce-core/blob/bfb6880afb4fbb70110f5cd8252af69806a104d9/src/main/java/io/lettuce/core/cluster/RedisClusterClient.java)

This implementation focuses on clusterClient.
 But on thinking more i think we can remove dependency on the client and make [RedisPlanStatisticsProvider](https://github.com/prestodb/presto/pull/20455/files#diff-1f960ef7009cb29e42148067ecc2ae4715b6f594d1ce037e27c7455e789400eb) depend on the **RedisAsyncCommands<String, HistoricalPlanStatistics>** interface instead.

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java (hunk 6 above)
Comment (jaystarshot) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java:59]: I based this off [com.facebook.presto.hive.HiveMetadataUpdateHandleThriftSerde#thriftCodecManager](https://github.com/prestodb/presto/blob/7805bf0d451fe336afa77827e0697ab7ac2353ee/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdateHandleThriftSerde.java#L50). Not sure if that needs to be initialized at each call but considered it safer to follow precedent

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisPlanStatisticsProvider.java (hunk 13 above)
Comment (jaystarshot) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisPlanStatisticsProvider.java]: Please re-review, I updated the initialization flow

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+            log.info("Redis Provider Plugin returned an empty instance");
+            return ImmutableList.of(EmptyPlanStatisticsProvider.getInstance());
+        }
+        if (!(propertyMap.getOrDefault(RedisProviderConfig.COORDINATOR_PROPERTY_NAME, "false").equals("true") &&
+                propertyMap.getOrDefault(RedisProviderConfig.HBO_PROVIDER_ENABLED_NAME, "false").equals("true"))) {
+            return ImmutableList.of(EmptyPlanStatisticsProvider.getInstance());
+        }
+        Injector injector = intializeAndGetInjector();
+        // Create Provider
+        RedisPlanStatisticsProvider redisPlanStatisticsProvider =
```
[the remark this answers is not in the document]
Comment (jaystarshot) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java]: Please re-review, I updated the initialization flow

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java (hunk 14 above)
Comment (jaystarshot) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java:49]: Please re-review, I updated the initialization flow

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+The following configuration properties are available:
+
+====================================== =====================================================================
+Property Name                          Description
+====================================== =====================================================================
+``coordinator``                        Boolean property whether Presto server is a coordinator
+``hbo.provider.server_uri``            Redis Server URI
+``hbo.provider.total-fetch-timeoutms`` Maximum timeout in ms for redis fetch requests
+``hbo.provider.total-set-timeoutms``   Maximum timeout in ms for redis set requests
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:25]: "Redis"

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+``hbo.provider.default-ttl-seconds``   TTL in seconds of the redis data to be stored
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:26]: "Redis"

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+``hbo.provider.enabled``               Boolean property whether this plugin is enabled in production
+``credentials-path``                   Path for redis credentials
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:28]: "Redis"

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+``hbo.provider.cluster-mode-enabled``  Boolean property whether cluster mode is enabled
+====================================== =====================================================================
+
+Credentials
+-----------
+
+The plugin requires the Redis Server URI property ``hbo.provider.server_uri`` in order to access Redis
+Based on your custom redis deployment you may need to add additional credentials
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:36]: "Redis"

File: presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+    <dependency>
+        <groupId>com.facebook.presto</groupId>
+        <artifactId>presto-hbo-provider</artifactId>
+        <version>${project.version}</version>
+        <type>zip</type>
+        <scope>provided</scope>
+    </dependency>
+
+4. You can add your custom redis client connection login in com.facebook.presto.statistic.RedisClusterAsyncCommandsFactory, just make sure that the AsyncCommands are provided properly
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/plugin/redis-hbo-provider.rst:83]: "Redis"

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+    private static final ThriftCodecManager thriftCodecManager = new ThriftCodecManager();
+
+    @Override
+    public String decodeKey(ByteBuffer bytes)
+    {
+        if (bytes.hasArray()) {
+            return StandardCharsets.UTF_8.decode(bytes).toString();
+        }
+        else {
+            throw new RuntimeException("Error decoding key");
```
Comment (tdcmeehan) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/HistoricalStatisticsSerde.java:46]: Maybe put something more descriptive (it seems that this shouldn't happen in practice)

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderConfig.java [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+
+    private long defaultTtlSeconds;
+    private boolean hboProviderEnabled;
+
+    private boolean clusterModeEnabled;
+    private long totalSetTimeoutMs;
+    private String credentialsPath;
+    private String redisPropertiesPath;
+
+    @Config("hbo.provider.cluster-mode-enabled")
```
Comment (tdcmeehan) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderConfig.java:34]: Please update the naming of all properties to include Redis HBO, e.g. 

`@Config("hbo.redis-provider.cluster-mode-enabled")`

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderConfig.java [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+        return this;
+    }
+
+    public String getRedisPropertiesPath()
+    {
+        return redisPropertiesPath;
+    }
+
+    // Constants
+    public static final String COORDINATOR_PROPERTY_NAME = "coordinator";
```
Comment (tdcmeehan) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderConfig.java:167]: Please move constants to the top

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/PropertiesUtil.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+        try {
+            properties = new HashMap<>(loadProperties(redisProviderFile));
+            // load secrets
+            File credentialFile = new File(properties.get(RedisProviderConfig.REDIS_CREDENTIALS_PATH));
+            if (credentialFile.exists()) {
+                properties.putAll(loadProperties(credentialFile));
+            }
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (tdcmeehan) [inline redis-hbo-provider/src/main/java/com/facebook/presto/statistic/PropertiesUtil.java:60]: Consider having it throw, rather than repackaging as `RuntimeException`.  If you're doing this because it's used in a constructor, you can move that code into a method that's annotated as `@PostConstruct`--that will ensure it's called after the constructor, but before runtime.  See examples in the code.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Mostly LGTM, some minor nits

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/PropertiesUtil.java (hunk 24 above)
Comment (jaystarshot) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/PropertiesUtil.java:60]: On testing, I think **@PostConstruct** only applies to container-managed beans and does not work here.

**Comment by jaystarshot** [comment]
@tdcmeehan  I think ` @PostConstruct` only applies to container-managed beans and except that I addressed your review comments. I also felt like the initializations were missing some tests so added a small one at [here](https://github.com/prestodb/presto/blob/9d26f6aa5331330b1b5d8b0aee9acbde0c47d618/redis-hbo-provider/src/test/java/com/facebook/presto/statistic/TestRedisProviderPlugin.java)

**Review by tdcmeehan — COMMENTED** [review]
See modified suggestion on what to do with the config properties.

File: redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+    public Iterable<HistoryBasedPlanStatisticsProvider> getHistoryBasedPlanStatisticsProviders()
+    {
+        if (this.propertyMap == null) {
+            log.info("Redis Provider Plugin returned an empty instance");
+            return ImmutableList.of(EmptyPlanStatisticsProvider.getInstance());
+        }
+        if (!(propertyMap.getOrDefault(RedisProviderConfig.COORDINATOR_PROPERTY_NAME, "false").equals("true") &&
+                propertyMap.getOrDefault(RedisProviderConfig.HBO_PROVIDER_ENABLED_NAME, "false").equals("true"))) {
+            return ImmutableList.of(EmptyPlanStatisticsProvider.getInstance());
+        }
```
[the remark this answers is not in the document]
Comment (jaystarshot) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java:66]: I agree yes I can do this and that would remove the constants and their use outside the config. 
 To confirm, we'd still have to manually load the .properties file from the storage because airlift lacks the functionality to do so natively. In other connectors we do the initialization through [ConnectorFactory](https://github.com/prestodb/presto/blob/cea8ab195d67cc798a0405b6c7a21edcd4aa9722/presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorFactory.java#L26) which gives the propertyMap directly. For plugins this is not supported. 
Hence I've modeled the loading process after [PrestoSparkLauncherCommand](https://github.com/prestodb/presto/blob/dd5ae3bd7a572f1c66d25b303af49fb3c2ab0802/presto-spark-launcher/src/main/java/com/facebook/presto/spark/launcher/PrestoSparkLauncherCommand.java#L68) .
Not sure if there is any other way

[the remark this answers is not in the document]
Comment (jaystarshot) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java:66]: > // double check: why do we want to check if coordinator == true? 
We do not want the workers to maintain an active connection to redis since only the co-ordinator is going to use them

[the remark this answers is not in the document]
Comment (jaystarshot) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java:66]: We also make use of a number of @Provides annotations within 
 [InjectorFactory](https://github.com/prestodb/presto/blob/73687d27a5495c2ed282bee1868c759180a7fe0d/redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderInjectorFactory.java). Managing them cleanly is something I'm unsure about. Additionally, since we don't need any classes if these configurations are set to false, we'll need to implement this check in the initialization of all classes as well as in each provision. An alternative approach would be to construct the configuration initially in a separate bootstrap instance and then utilize those configuration variables. Alternatively, we could simply leave the code as it is if we must load the file anyway.

[the remark this answers is not in the document]
Comment (tdcmeehan) [reply redis-hbo-provider/src/main/java/com/facebook/presto/statistic/RedisProviderPlugin.java:66]: I see the problem.  Let's keep it as is.

**Comment by jaystarshot** [comment]
@tdcmeehan  Thanks a lot for the detailed review! After rebase all tests except product-tests-basic-environment  seems to have passed and that test error logs seems [flaky](https://github.com/prestodb/presto/actions/runs/5835238982/job/15826438881?pr=20455).  Can we merge without that or can you please rerun or grant me rerun permission?

File: presto-main/etc/config.properties [hunk 26]
```diff
@@ -27,6 +27,7 @@ query.client.timeout=5m
 query.min-expire-age=30m
 
 plugin.bundles=\
+   ../redis-hbo-provider/pom.xml,\
```
Comment (jaystarshot) [inline presto-main/etc/config.properties:30]: Removing plugin from loading during presto dev intializations

**Comment by tdcmeehan** [comment]
@jaystarshot I added you as a contributor to the PrestoDB repo.  This should give you access to rerun tests and also label and assign reviewers.  Thanks for this contribution!

**Comment by jaystarshot** [comment]
Thank you!
